File: backend/app/schemas/ai_analysis.py
"""AI 시황 분석 및 포트폴리오 추천 스키마."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel


class MarketIndexItem(BaseModel):
    symbol: str
    name: str
    price: float | None
    change_pct: float | None
    week_change_pct: float | None


class ExchangeRateInfo(BaseModel):
    usd_krw: float | None
    change_pct: float | None


class SectorInfo(BaseModel):
    sector: str
    etf_ticker: str
    change_pct: float | None


class PortfolioRisk(BaseModel):
    score: int
    concentration_risk: str
    sector_bias: list[str]
    description: str


class RecommendedAction(BaseModel):
    ticker: str
    name: str
    action: str
    reason: str
    priority: str


class AlternativePortfolioItem(BaseModel):
    ticker: str
    name: str
    weight: float
    reason: str


class AlternativePortfolio(BaseModel):
    risk_level: str
    expected_return: str
    items: list[AlternativePortfolioItem]


class AIAnalysisResult(BaseModel):
    market_summary: str
    portfolio_risk: PortfolioRisk
    recommendations: list[RecommendedAction]
    alternative_portfolios: list[AlternativePortfolio]
    disclaimer: str


class AIAnalysisResponse(BaseModel):
    status: str
    cached_at: datetime | None
    market_indices: list[MarketIndexItem]
    exchange_rate: ExchangeRateInfo
    sector_performance: list[SectorInfo]
    analysis: AIAnalysisResult | None
    error_message: str | None

File: frontend/src/api/aiAnalysis.ts
import { api } from "@/api/client";

export interface MarketIndexItem {
  symbol: string;
  name: string;
  price: number | null;
  change_pct: number | null;
  week_change_pct: number | null;
}

export interface ExchangeRateInfo {
  usd_krw: number | null;
  change_pct: number | null;
}

export interface SectorInfo {
  sector: string;
  etf_ticker: string;
  change_pct: number | null;
}

export interface PortfolioRisk {
  score: number;
  concentration_risk: string;
  sector_bias: string[];
  description: string;
}

export interface RecommendedAction {
  ticker: string;
  name: string;
  action: string;
  reason: string;
  priority: string;
}

export interface AlternativePortfolioItem {
  ticker: string;
  name: string;
  weight: number;
  reason: string;
}

export interface AlternativePortfolio {
  risk_level: string;
  expected_return: string;
  items: AlternativePortfolioItem[];
}

export interface AIAnalysisResult {
  market_summary: string;
  portfolio_risk: PortfolioRisk;
  recommendations: RecommendedAction[];
  alternative_portfolios: AlternativePortfolio[];
  disclaimer: string;
}

export interface AIAnalysisResponse {
  status: "ready" | "error";
  cached_at: string | null;
  market_indices: MarketIndexItem[];
  exchange_rate: ExchangeRateInfo;
  sector_performance: SectorInfo[];
  analysis: AIAnalysisResult | null;
  error_message: string | null;
}

export const fetchAIAnalysis = (forceRefresh = false) =>
  api
    .get<AIAnalysisResponse>("/ai-analysis", { params: { force_refresh: forceRefresh } })
    .then((r) => r.data);
